be/src/column/datum_convert.cpp

```diff
@@ -145,4 +147,314 @@ std::string datum_to_string(TypeInfo* type_info, const Datum& datum) {
+    if (src_datum.is_null()) {
+        dst_datum.set_null();
+        return Status::OK();
+    }
+
+    if (src_type == OLAP_FIELD_TYPE_DECIMAL_V2) {
+        decimal12_t dst_value(src_datum.get_decimal().int_value(), src_datum.get_decimal().frac_value());
+        dst_datum.set_decimal12(dst_value);
+        return Status::OK();
+    }
+
+    if (src_type == OLAP_FIELD_TYPE_DECIMAL128) {
+        const auto& src_value = src_datum.get_decimal();
+        decimal12_t tmp;
+        tmp.integer = src_value.int_value();
+        tmp.fraction = src_value.frac_value();
+        dst_datum.set_decimal12(tmp);
+        return Status::OK();
+    }
+
+    std::stringstream ss;
+    ss << "does not support change " << field_type_to_string(src_type) << " to " << field_type_to_string(dst_type);
+    return Status::InternalError(ss.str());
+}
+
+Status convert_to_decimal_v2(FieldType src_type, const Datum& src_datum, FieldType dst_type, Datum& dst_datum) {
+    if (src_datum.is_null()) {
+        dst_datum.set_null();
+        return Status::OK();
+    }
+
+    if (src_type == OLAP_FIELD_TYPE_DECIMAL) {
+        DecimalV2Value dst_value;
+        dst_value.from_olap_decimal(src_datum.get_decimal12().integer, src_datum.get_decimal12().fraction);
+        dst_datum.set_decimal(dst_value);
+        return Status::OK();
+    }
+
+    if (src_type == OLAP_FIELD_TYPE_DECIMAL128) {
+        auto src_value = src_datum.get_int128();
+        DecimalV2Value dst_value;
+        strings::memcpy_inlined(&dst_value, &src_value, sizeof(int128_t));
+        dst_datum.set_decimal(dst_value);
+    }
+
+    std::stringstream ss;
+    ss << "does not support change " << field_type_to_string(src_type) << " to " << field_type_to_string(dst_type);
+    return Status::InternalError(ss.str());
+}
+
+Status convert_to_decimal128(FieldType src_type, const Datum& src_datum, FieldType dst_type, Datum& dst_datum) {
+    if (src_datum.is_null()) {
+        dst_datum.set_null();
+        return Status::OK();
+    }
+
+    if (src_type == OLAP_FIELD_TYPE_DECIMAL) {
+        auto src_value = src_datum.get_decimal12();
+        DecimalV2Value dst_value;
+        dst_value.from_olap_decimal(src_value.integer, src_value.fraction);
+        dst_datum.set_int128(dst_value);
+        return Status::OK();
+    }
+
+    if (src_type == OLAP_FIELD_TYPE_DECIMAL_V2) {
+        auto src_value = src_datum.get_decimal();
+        int128_t dst_value;
+        strings::memcpy_inlined(&dst_value, &src_value, sizeof(int128_t));
+        dst_datum.set_int128(dst_value);
+        return Status::OK();
+    }
+
+    std::stringstream ss;
+    ss << "does not support change " << field_type_to_string(src_type) << " to " << field_type_to_string(dst_type);
+    return Status::InternalError(ss.str());
+}
+
+Status convert_datum(TypeInfo* src_type_info, const Datum& src_datum, TypeInfo* dst_type_info, Datum& dst_datum) {
```

Comment:
I think a separate conversion function could be written here for each of the two types. This would avoid having if else for each conversion and would make full use of vectorization

Reply:
Seems that it's true. I will fix it in next commit

Reply:
fixed in https://github.com/StarRocks/starrocks/pull/453/commits/b2240b204328e02b6113d118d70b473fcada4f68, PTAL

Reply:
> I think a separate conversion function could be written here for each of the two types. This would avoid having if else for each conversion and would make full use of vectorization

fixed in https://github.com/StarRocks/starrocks/pull/453/commits/b2240b204328e02b6113d118d70b473fcada4f68, PTAL